addon/model-data.js

```diff
@@ -0,0 +1,402 @@
+    this._attributes[key] = value;
+
+    if (key in this._inFlightAttributes) {
+      originalValue = this._inFlightAttributes[key];
+    } else {
+      originalValue = this._data[key];
+    }
+    // If we went back to our original value, we shouldn't keep the attribute around anymore
+    if (value === originalValue) {
+      delete this._attributes[key];
+    }
+  }
+
+  getAttr(key) {
+    if (key in this._attributes) {
+      return this._attributes[key];
+    } else if (key in this._inFlightAttributes) {
+      return this._inFlightAttributes[key];
+    } else {
+      return this._data[key];
+    }
+  }
+
+  hasAttr(key) {
+    return key in this._attributes || key in this._inFlightAttributes || key in this._data;
+  }
+
+  unloadRecord() {
+    if (this.isDestroyed) {
+      return;
+    }
+    this.reset();
+    this.destroy();
+  }
+
+  destroy() {
+    this.isDestroyed = true;
+    this.storeWrapper.disconnectRecord(this.modelName, this.id, this.clientId);
+  }
+
+  removeFromInverseRelationships() {}
+
+  clientDidCreate() {}
+
+  getOrCreateNestedModelData(key, modelName, id, internalModel) {
+    let nestedModelData = this._nestedModelDatas[key];
+    if (!nestedModelData) {
+      nestedModelData = this._nestedModelDatas[key] = this.createNestedModelData(
+        modelName,
+        id,
+        internalModel
+      );
+    }
+    return nestedModelData;
+  }
+
+  createNestedModelData(modelName, id, internalModel) {
+    let storeWrapper = new NestedModelDataWrapper(internalModel);
+    return new M3ModelData(modelName, id, null, storeWrapper, this.store);
+  }
+
+  destroyNestedModelData(key) {
+    let nestedModelData = this._nestedModelDatas[key];
+    if (nestedModelData) {
+      // destroy
+      delete this._nestedModelDatas[key];
+    }
+  }
+
+  hasNestedModelData(key) {
+    return !!this._nestedModelDatas[key];
+  }
+
+  get _attributes() {
+    if (this.__attributes === null) {
+      this.__attributes = Object.create(null);
+    }
+    return this.__attributes;
+  }
+
+  set _attributes(v) {
+    this.__attributes = v;
+  }
+
+  get _data() {
+    if (this.__data === null) {
+      this.__data = Object.create(null);
+    }
+    return this.__data;
+  }
+
+  set _data(v) {
+    this.__data = v;
+  }
+
+  get _inFlightAttributes() {
+    if (this.__inFlightAttributes === null) {
+      this.__inFlightAttributes = Object.create(null);
+    }
+    return this.__inFlightAttributes;
+  }
+
+  set _inFlightAttributes(v) {
+    this.__inFlightAttributes = v;
+  }
+
+  get _nestedModelDatas() {
+    if (this.__nestedModelsData === null) {
+      this.__nestedModelsData = Object.create(null);
+    }
+    return this.__nestedModelsData;
+  }
+
+  /**
+   *
+   * @param updates
+   * @param nestedCallback a callback for updating the data of a nested model-data instance
+   * @returns {Array}
+   * @private
+   */
+  _mergeUpdates(updates, nestedCallback, calculateChanges = true) {
```

Comment:
@hjdivad I think it's because of the usage of this function in `didCommit` function, where we don't receive the `calculateChanges` param. However, if the `_mergeUpdates` is called, we definitely want to calculate changes, because we only do it if `data` was specified.

I think the whole thing stems from the fact that `didCommit` "cannot" be called without backing record, but it is something, which might be worth fixing in ED.